### ISSUE TITLE
Annotation-Based API for Consuming Messages

### ISSUE THREAD

**Issue by dblevins** [eclipse-ee4j/jms-api#243]
TODO

**Comment by hantsy** [comment]
Spring can process jms via simple `@JmsListener`, how hard we still need the interface?

```java

@ApplicationScoped 
@Slf4j
public class Receiver {

    @JmsListener(destination = "hello")
    public void onMessage(String message) {
        log.debug("receving message: {}", message);
    }

}
```

Provides programmatic APIs to set the JSM global properties, and also allow to set `properties` attribute(a `Map` or a string of properties joint by ",") on the `JmsListener` annotation.

**Comment by OndroMih** [comment]
I agree with @hantsy, no interface is needed. And I also like that any CDI bean can be turned into a JMS observer. However, I'm not sure whether a single @JmsListener is enough to distinguish between a queue and a topic. An implementation might need to know whether the listener wants to connect to a queue or a topic.

Another idea is to turn this into a native CDI event observer, with a @JmsListener qualifier. Then all the CDI event API could be used, including asynchronous API. Using `@JmsListener` on the method could be also allowed but would work as a shorthand, e.g.:

```
@JmsListener(destination = "hello")
    public void onMessage(String message) {
        log.debug("receving message: {}", message);
    }
```

would be equivalent to:

```
    public void onMessage(@Observes @JmsListener(destination = "hello") String message) {
        log.debug("receving message: {}", message);
    }
```

The latter syntax also allows injection per method call, e.g. the following would inject a myProcessor CDI bean:

```
    public void onMessage(@Observes @JmsListener(destination = "hello") String message, MyProcessor myProcessor) {
        myProcessor.process(message);
    }
```

We could allow injecting some contextual information, if it's technically possible. E.g. inject a CDI bean of type of Topic/Queue for the current queue/topic. But we would need to explore that separately, I'm not sure if CDI allows this dynamically for each message.

**Comment by m-reza-rahman** [comment]
This is a very deep topic with a number of discussions that took place over a long time but never implemented. If the idea is to finally do this work, I suggest starting with a very simple straw man and detailed follow up discussions on the mailing list. I would be delighted to participate and share incrementally all the things from the past. It’s too much for this one issue I think. The interim outcome may be creating several smaller issues based on some initial progress that hopefully can go into Jakarta EE 11.

**Comment by hantsy** [comment]
@m-reza-rahman I remember JBoss Seam2/3(Seam 3 is based CDI) has very simple approach to bridge the JMS to CDI event observer, this JMS handling should be simple as possible.

**Comment by m-reza-rahman** [comment]
For sake of completeness, I'll share the same code example here that I shared with the JMS 2 specification lead Nigel Deakin around 2014 on this topic:

```java
@ApplicationScoped
@MaxConcurrency(10)
public class HandlingEventRegistrationAttemptConsumer {
    @MessageListener(
        destinationLookup="jms/HandlingEventRegistrationAttemptQueue",
        selector="source = 'mobile'",
        batchSize=10, orderBy=TIMESTAMP,
        retry=5, retryDelay=7000, deadLetterQueue="jms/DeadLetterQueue")
    public void onEventRegistrationAttempt(
        HandlingEventRegistrationAttempt... attempts) {
        ...
    }
}
```

**Comment by hantsy** [comment]
To make it as simple as possible, Spring has already provided good examples of the usage of JMS. 

https://github.com/jakartaee/messaging/issues/259#issuecomment-1837332992
